### PR TITLE
Typing for ``./pylint``, batch 1

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Similar to #4950 but now for `./pylint`, going directory for directory.
For this PR I would like to do 3 directories:
- [x] `./pylint/utils`
- [x] `./pylint/testutils`
- [x] `./pylint/reporters`

### CODE REVIEW

File: pylint/checkers/similar.py
Comment (DanielNoord) [inline pylint/checkers/similar.py:378]: We might want to refactor this, not the best of typing..

File: pylint/utils/ast_walker.py
Comment (DanielNoord) [inline pylint/utils/ast_walker.py:18]: This is missing more specific types, will fix in later commit

File: pylint/utils/utils.py
Comment (DanielNoord) [inline pylint/utils/utils.py:68]: Is this `nodes.NodeNG`?

**Comment by cdce8p** [comment]
I would suggest to wait until after at least #4940 is merged.

File: pylint/testutils/output_line.py
Comment (DanielNoord) [inline pylint/testutils/output_line.py:23]: I could not get this to work... Perhaps somebody else wants to take a look at this issue and how to solve it!

**Comment by DanielNoord** [comment]
Tests are failing on the fact that `re.Pattern` is not available in Python 3.6.
Is there a known workaround for this? On two places I use `isinstance(x, re.Pattern)`. Can that be overwritten by something like `isinstance(x, re.BaseClass)` which is available in 3.6?

**Comment by Pierre-Sassoulas** [comment]
> Is there a known workaround for this? On two places I use isinstance(x, re.Pattern). Can that be overwritten by something like isinstance(x, re.BaseClass) which is available in 3.6?

We could wait until the 23rd of December 2021 and drop python 3.6 before adding this typing and removing the ``type: ignore`` ? 😄 Might not be worth our time to fix this issue in 3.6 as very soon we won't have to.

**Comment by DanielNoord** [comment]
I'll do that when I get back to my computer. Need to wait for #4940 anyway.

**Comment by DanielNoord** [comment]
Any idea why the docs build is failing? Should I add `pytest` to the dependencies?

**Comment by cdce8p** [comment]
> Any idea why the docs build is failing? Should I add `pytest` to the dependencies?

Better to guard the `pytest` import behind `if TYPE_CHECKING` and to use strings for the imported types (from pytest).

File: pylint/checkers/similar.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/similar.py:378]: I agree, but maybe in another merge request ?

File: pylint/lint/pylinter.py
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:734]:
```suggestion
            and self.stats.get("by_msg") is not None
```
Not doing that create nasty error if `by_msg == ""` or ``by_msg == 0``

File: pylint/reporters/reports_handler_mix_in.py
Comment (Pierre-Sassoulas) [inline pylint/reporters/reports_handler_mix_in.py:64]:
```suggestion
        if not isinstance(self, BaseChecker) or not isinstance(self, ReportsHandlerMixIn):
            raise TypeError
```
Premature return so the following code is simpler to understand with less to keep in mind and less indented.

File: pylint/reporters/text.py
Comment (Pierre-Sassoulas) [inline pylint/reporters/text.py:67]:
```suggestion
def _get_ansi_code(color: Optional[str] = None, style: Optional[str] = None) -> str:
```

File: pylint/testutils/output_line.py
Comment (Pierre-Sassoulas) [inline pylint/testutils/output_line.py:97]: Something we might fix if we do #1590 (By uniformizing all the functional test results)

File: pylint/checkers/base.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/base.py:1162]: In tests we can handle it this way, but we  can't do this in pylint's code. We probably need to remove the possibility that it's not an int altogether by doing a refactor. Crashing with an assert will still be a crash.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py:1162]: Are you sure? Based on the code on L1163 I think `self.stats["module"]` will always be an `int`. Otherwise it would already produce crashes (and none seem to be reported).
I think this actually asserts something that is already always the case but is needed for `mypy` not to fail (and as a helpful remainder for us that this stat should always be an `int`).

**Comment by DanielNoord** [comment]
There are some `type: ignore`'s in this final version. However, I would be in favour of merging this and adding typing to all other folders and then get back to those comments.
Some of those require extensive refactoring for which I currently do not have the time.

File: pylint/checkers/base.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/base.py:1162]: It makes mypy stop warning about the issue but it's the wrong way to fix the issue imo.
- It lower performance because we'll now make an assert at run time
- The root cause of ``self.stats`` not being always properly initialized and ``visit_module`` not being safely callable in badly initialized child class is not addressed (default value for ``self.stats`` is None).
- The error is only marginally better as we'd have a ``TypeError: can only concatenate str (not "int") to str`` already or something similarly clear.
- This is not the case here, but in general for pylint this kind of issue could come from not anticipating a strange code construct that will come from astroid. For example we think we can have only an inner function, but it happens the user defined an inner class. Asserting that we have a ``FunctionDef`` when we have a ``ClassDef`` instead will still crash pylint the same way that if we did nothing at all. It's just silencing the linter with an assumption we made.

I think this is a design issue on the base checker that would require a refactor and we may as well just ignore it until we do the refactor.

File: pylint/extensions/docparams.py
Comment (DanielNoord) [inline pylint/extensions/docparams.py:226]: This is a similar issue to the `assert` I have just removed (due to `linter.stats` having multiple types). Is this okay? Or do you want me to add `type: ignore` here as well?

File: pylint/extensions/typing.py
Comment (DanielNoord) [inline pylint/extensions/typing.py:143]: Same here as above.

File: pylint/extensions/docparams.py
Comment (Pierre-Sassoulas) [reply pylint/extensions/docparams.py:226]: I think this could be checked once when parsing the option, so we can also ignore.

Comment (DanielNoord) [reply pylint/extensions/docparams.py]: Done!

File: pylint/testutils/lint_module_test.py
Comment (cdce8p) [inline pylint/testutils/lint_module_test.py:141]: `IOBase` or `TextIOBase`?

File: pylint/testutils/lint_module_test.py
Comment (cdce8p) [inline pylint/testutils/lint_module_test.py:169]: The annotation should not be necessary

File: pylint/testutils/output_line.py
Comment (cdce8p) [inline pylint/testutils/output_line.py:97]: Couldn't it fail if `row` is a string of length `6`. `isinstance(row, tuple)` might work

File: pylint/testutils/output_line.py
Comment (cdce8p) [inline pylint/testutils/output_line.py:105]:
```suggestion
            raise
```
That would be enough, but why add the `isinstance` check in the first place?

File: pylint/testutils/reporter_for_tests.py
Comment (cdce8p) [inline pylint/testutils/reporter_for_tests.py:52]: Either this is always true, in which case it shouldn't be there in the first place or `result` is possibly unbound.
For 1: Add `out: StringIO` below the class docstring.

File: pylint/utils/ast_walker.py
Comment (cdce8p) [inline pylint/utils/ast_walker.py:31]: `BaseChecker`?

File: pylint/utils/ast_walker.py
Comment (cdce8p) [inline pylint/utils/ast_walker.py:61]: Probably `nodes.NodeNG`

File: pylint/utils/file_state.py
Comment (cdce8p) [inline pylint/utils/file_state.py:134]: Please don't remove argument names even if unused

File: pylint/utils/utils.py
Comment (cdce8p) [inline pylint/utils/utils.py:130]: `IOBase`

File: pylint/utils/utils.py
Comment (cdce8p) [inline pylint/utils/utils.py:133]: Can't you use a base class here? `codecs.StreamReader`?

File: pylint/utils/utils.py
Comment (cdce8p) [inline pylint/utils/utils.py:292]:
```suggestion
        value = cast(Pattern, value).pattern
```

File: tests/test_self.py
Comment (cdce8p) [inline tests/test_self.py:123]: Properties are usually not an issue for mypy.
`IOBase`?

File: tests/test_self.py
Comment (cdce8p) [inline tests/test_self.py:127]: Same here with the `type: ignore` comments

File: tests/utils/unittest_ast_walker.py
Comment (cdce8p) [inline tests/utils/unittest_ast_walker.py:45]: The issue is that `MockLinter` doesn't extend `PyLinter`

**Review by cdce8p — CHANGES_REQUESTED** [review]
Thanks for your work, but tbh this was really difficult to review. It would be better if you could limit typing changes to much smaller PRs in the future.

One thing that stood out to me was `Union[BufferedReader, BytesIO, TextIO]`. It might be better to use a common base class where possible. Something like `IOBase`. I annotated it a bunch of times, haven't checked it in details though.

--
One last comment. To make the review process a bit easier, I would appreciate if you don't resolve any of the conversations. I won't be able to find comments otherwise. If you agree, simply add a 👍🏻 and change the code, I'll later go through and resolve them. If you like to start a conversation on a resolved item, please unresolve it.

File: pylint/checkers/similar.py
Comment (DanielNoord) [reply pylint/checkers/similar.py:382]: I thought adding typing would be better so we can start moving to adding more mypy flags such as `--disallow-untyped-defs` etc.
I can add a `type: ignore` if you want, but personally I like how `Any` typings serve as a visual reminder that something is left to be done and allow searching for them within the project directory to create a "typing to-do list".

File: pylint/extensions/code_style.py
Comment (DanielNoord) [reply pylint/extensions/code_style.py:85]: Oops.. 😄

File: pylint/reporters/multi_reporter.py
Comment (DanielNoord) [reply pylint/reporters/multi_reporter.py:92]: This is based on what the current test suite does. Apparently this function never gets called with `None` or `str`.
However, I think it would be better to change this indeed. I'm going to change this to `Optional[PosixPath, str]` and do a separate commit in which I do `Union[None, PosixPath, str]` -> `Optional[PosixPath, str]` for all other cases.

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [reply pylint/reporters/reports_handler_mix_in.py:67]: Rather have me put a `type: ignore` here?

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [reply pylint/reporters/reports_handler_mix_in.py:79]: The `type: ignore` is because `mypy` doesn't recognise `hasattr`, but I thought it's warning was fair and that we might want a check for it.

File: pylint/testutils/lint_module_test.py
Comment (DanielNoord) [reply pylint/testutils/lint_module_test.py:99]: `Mypy` complains about this. Want me to add `type ignore`?

File: pylint/testutils/lint_module_test.py
Comment (DanielNoord) [reply pylint/testutils/lint_module_test.py:141]: `TextIOBase`

File: pylint/testutils/lint_module_test.py
Comment (DanielNoord) [reply pylint/testutils/lint_module_test.py:169]: `Mypy` complains when it is missing. Want me to add an ignore?

File: pylint/testutils/output_line.py
Comment (DanielNoord) [reply pylint/testutils/output_line.py:97]: Sorry! Totally forgot about that!

File: pylint/testutils/output_line.py
Comment (DanielNoord) [reply pylint/testutils/output_line.py:105]: Because `MalformedOutputLineException` requires a `ValueError` and `mypy` complains otherwise.

File: tests/utils/unittest_ast_walker.py
Comment (DanielNoord) [reply tests/utils/unittest_ast_walker.py:45]: Yes, but I also can't type `MockLinter` here. What do you want me to do?

**Comment by DanielNoord** [comment]
I'm truly sorry for making this PR too large. I got a bit carried away when working and forgot to keep a close eye on how truly large this PR had become. If you want I can still split it up based on the three directories and created individual PR's based on your feedback?

I'm working on incorporating all of your feedback. One of the issues I ran into is that `TextIO` doesn't inherit from `IOBase`. It only inherits from `IO`. That class is often too broad for typing and would lead to additional `isinstance` checks to satisfy `mypy`. `TextIO` is returned by some of the functions in the `sys` library and is therefore also needed as a recurring type. Previously I solved this by adding the `Union`'s. I can see why you don't like that solution, but I'm stuck thinking of something better.
TLDR: `IO` creates `mypy` issues because of its broadness and the lack of certain standard attributes and methods of its sub-classes, `TextIO` doesn't inherit from `IOBase` but is returned by `sys` functions, current `Union`'s are too broad and ugly.

I'm sorry for resolving the conversation prematurely. I was used to resolving after incorporating feedback but I can see how this is unhelpful for you. I appreciate the effort you took to still review this PR instead of dismissing it because of its size (which I have seen happen in other projects). Thank you!

**Comment by cdce8p** [comment]
I haven't taken a look at all the review comments yet. Only want to answer some of your questions now.

> If you want I can still split it up based on the three directories and created individual PR's based on your feedback?

In general smaller PRs might be better. Don't know though if directory bases is the best approach. Although I know it will be more work, I would think of it in related topics. Just to name some examples:
* Everything dealing with `self.stats`
* `stream` and `IOBase`
* `get_global_option` and related `type: ignore` comments. (I would appreciate this one especially!)
* `filepath`
* ... (there might be more, but that's a start)

> One of the issues I ran into is that `TextIO` doesn't inherit from `IOBase`.

To be honest with you, I haven't checked it in detail if it would work / where it might not. (There were just too many changes.)
Separating those into a new PR might be a way to better understand the relationships between all of them. In the end, if `Union` is the only thing that works, so be it.

> I'm sorry for resolving the conversation prematurely. I was used to resolving after incorporating feedback but I can see how this is unhelpful for you.

It's nothing personal. With so many different conversations I just felt that otherwise I couldn't follow it properly.

> I appreciate the effort you took to still review this PR instead of dismissing it because of its size (which I have seen happen in other projects). Thank you!

I can understand why it sometimes happens. Most maintainers, I would guess, don't know every inch of their code of the top of their head. So it takes a lot of mental energy to go through it if at all.

**Review by Pierre-Sassoulas — APPROVED** [review]
Thanks ! This uncovered so many things that ideally we'd have to refactor, it's definitely a step in the right direction.

**Review by Pierre-Sassoulas — APPROVED** [review]
👍

**Review by cdce8p — APPROVED** [review]

**Pull request merged**
